Use Cython wrappers even in non-JIT code?

Currently we implement every primitive twice, once via `mpi4py` for use in non-jitted code, and once via Cython for jitted code. Instead, we could also replace the non-jit implementations with

```python
def send_impl(*args, **kwargs):
    return xla.apply_primitive(send_p, *args, **kwargs)
```

which would then always apply our Cython implementation.

#### Pro:
- Less code to maintain
- Consistent logging and debugging behavior (with `MPI4JAX_DEBUG`)
- Consistent user experience regarding array types and validation

#### Con:
- Our wrappers are more rudimentary than `mpi4py`s with less validation and limited errors
- No way to use `mpi4jax` without the Cython extensions

I was playing around with 
```python
def mpi_allreduce_impl(x, token, op, comm):
    return xla.apply_primitive(mpi_allreduce_p, token, op=op, comm=comm)
```

and it errors with

```phyhon
    def apply_primitive(prim, *args, **params):
      """Impl rule that compiles and runs a single primitive 'prim' using XLA."""
>     compiled_fun = xla_primitive_callable(prim, *unsafe_map(arg_spec, args), **params)
E     TypeError: unhashable type: 'mpi4py.MPI.Op'
```

I guess we must first encode the types

--- a/mpi4jax/collective_ops/allreduce.py
+++ b/mpi4jax/collective_ops/allreduce.py
@@ -102,7 +102,7 @@ def mpi_allreduce_abstract_eval(xs, token, op, comm):
 
 
 mpi_allreduce_p.multiple_results = True
-mpi_allreduce_p.def_impl(default_primitive_impl(mpi_allreduce_p))
+mpi_allreduce_p.def_impl(mpi_allreduce_impl)
 mpi_allreduce_p.def_abstract_eval(mpi_allreduce_abstract_eval)
 
 # assign to the primitive the correct encoder

--- a/mpi4jax/collective_ops/allreduce.py
+++ b/mpi4jax/collective_ops/allreduce.py
@@ -109,6 +109,8 @@ def mpi_allreduce_value_and_jvp(in_args, tan_args, op, comm):
     res = Allreduce(x, token=token, op=op, comm=comm)
 
     # Identify the correct adjoint
+    op = unpack_hashable(op)
+
     if op == _MPI.SUM:
         jvp = (x_tan, token_tan)
     else:

--- a/mpi4jax/collective_ops/allreduce.py
+++ b/mpi4jax/collective_ops/allreduce.py
@@ -17,12 +17,14 @@ from ..utils import (
     dtype_ptr,
     wrap_as_hashable,
     unpack_hashable,
+    default_primitive_impl,
 )
 
 from ..warn import warn_missing_omnistaging
 
 # The Jax primitive
 mpi_allreduce_p = Primitive("allreduce_mpi")  # Create the primitive
+mpi_allreduce_impl = default_primitive_impl(mpi_allreduce_p)
 
 
 # This function applies the primitive to an AST
@@ -30,14 +32,9 @@ def Allreduce(x, op, comm=_MPI.COMM_WORLD, token=None):
     if token is None:
         token = create_token(x)
 
-    return mpi_allreduce_p.bind(x, token, op=op, comm=comm)
-
-
-#  this function executes the primitive, when not under any transformation
-def mpi_allreduce_impl(x, token, op, comm):
     op = wrap_as_hashable(op)
     comm = wrap_as_hashable(comm)
-    return xla.apply_primitive(mpi_allreduce_p, x, token, op=op, comm=comm)
+    return mpi_allreduce_p.bind(x, token, op=op, comm=comm)
 
 
 #  This function compiles the operation
@@ -86,7 +83,7 @@ def mpi_allreduce_abstract_eval(xs, token, op, comm):
 
 
 mpi_allreduce_p.multiple_results = True
-mpi_allreduce_p.def_impl(mpi_allreduce_impl)
+mpi_allreduce_p.def_impl(default_primitive_impl(mpi_allreduce_p))
 mpi_allreduce_p.def_abstract_eval(mpi_allreduce_abstract_eval)
 
 # assign to the primitive the correct encoder

--- a/mpi4jax/collective_ops/allreduce.py
+++ b/mpi4jax/collective_ops/allreduce.py
@@ -2,8 +2,7 @@ import numpy as _np
 
 from mpi4py import MPI as _MPI
 
-from jax import abstract_arrays, device_put
-from jax import numpy as jnp
+from jax import abstract_arrays
 from jax.lax import create_token
 from jax.core import Primitive
 from jax.lib import xla_client
@@ -34,21 +33,9 @@ def Allreduce(x, op, comm=_MPI.COMM_WORLD, token=None):
 
 #  this function executes the primitive, when not under any transformation
 def mpi_allreduce_impl(x, token, op, comm):
-    # TODO: make this support gpus (use cupy?)
-    inpt = _np.asarray(x)
-    out = _np.zeros_like(x)
-
-    comm.Allreduce(inpt, out, op=op)
-
-    res = jnp.array(out, dtype=inpt.dtype)
-
-    # if it's a jax array and not a standard python array
-    if hasattr(x, "device_buffer"):
-        # put the result on the correct device if needed
-        if not (res.device_buffer.device() == x.device_buffer.device()):
-            res = device_put(res, device=x.device_buffer.device())
-
-    return res, token
+    hashableop = type('hop', (type(op),), dict(__hash__=lambda x: id(op)))
+    # op.__hash__ = lambda: id(op)
+    return xla.apply_primitive(mpi_allreduce_p, x, token, op=hashableop(), comm=comm)
 
 
 #  This function compiles the operation

--- a/mpi4jax/collective_ops/allreduce.py
+++ b/mpi4jax/collective_ops/allreduce.py
@@ -15,6 +15,8 @@ from ..utils import (
     _constant_s32_scalar,
     _constant_u64_scalar,
     dtype_ptr,
+    wrap_as_hashable,
+    unpack_hashable,
 )
 
 from ..warn import warn_missing_omnistaging
@@ -33,14 +35,17 @@ def Allreduce(x, op, comm=_MPI.COMM_WORLD, token=None):
 
 #  this function executes the primitive, when not under any transformation
 def mpi_allreduce_impl(x, token, op, comm):
-    hashableop = type('hop', (type(op),), dict(__hash__=lambda x: id(op)))
-    # op.__hash__ = lambda: id(op)
-    return xla.apply_primitive(mpi_allreduce_p, x, token, op=hashableop(), comm=comm)
+    op = wrap_as_hashable(op)
+    comm = wrap_as_hashable(comm)
+    return xla.apply_primitive(mpi_allreduce_p, x, token, op=op, comm=comm)
 
 
 #  This function compiles the operation
 def mpi_allreduce_xla_encode(c, x, token, op, comm):
     warn_missing_omnistaging()
+
+    op = unpack_hashable(op)
+    comm = unpack_hashable(comm)
 
     c = _unpack_builder(c)
     x_shape = c.GetShape(x)

--- a/mpi4jax/collective_ops/recv.py
+++ b/mpi4jax/collective_ops/recv.py
@@ -2,8 +2,7 @@ import numpy as _np
 
 from mpi4py import MPI as _MPI
 
-import jax.numpy as jnp
-from jax import abstract_arrays, device_put
+from jax import abstract_arrays
 from jax.lax import create_token
 from jax.core import Primitive
 from jax.lib import xla_client
@@ -41,20 +40,8 @@ def Recv(
 
 
 #  this function executes the primitive, when not under any transformation
-def mpi_recv_impl(x, token, source, tag, comm, status):
-    # TODO: make this support gpus (use cupy?)
-    out = _np.empty_like(x)
-    comm.Recv(out, source=source, tag=tag, status=status)
-
-    res = jnp.array(out, dtype=out.dtype)
-
-    # if it's a jax array and not a standard python array
-    if hasattr(x, "device_buffer"):
-        # put the result on the correct device if needed
-        if not (res.device_buffer.device() == x.device_buffer.device()):
-            res = device_put(res, device=x.device_buffer.device())
-
-    return res, token
+def mpi_recv_impl(*args, **kwargs):
+    return xla.apply_primitive(mpi_recv_impl, *args, **kwargs)
 
 
 #  This function compiles the operation
@@ -73,7 +60,7 @@ def mpi_recv_xla_encode(c, x, token, source, tag, comm, status):
     _dtype_ptr = dtype_ptr(dtype)
 
     sh = xla_client.Shape.tuple_shape(
-        [xla_client.Shape.array_shape(dtype, dims), xla_client.Shape.token_shape(),]
+        [xla_client.Shape.array_shape(dtype, dims), xla_client.Shape.token_shape()]
     )
 
     if status is None:

--- a/mpi4jax/collective_ops/recv.py
+++ b/mpi4jax/collective_ops/recv.py
@@ -15,6 +15,8 @@ from ..utils import (
     _constant_s32_scalar,
     _constant_u64_scalar,
     dtype_ptr,
+    wrap_as_hashable,
+    unpack_hashable,
 )
 
 from ..warn import warn_missing_omnistaging
@@ -40,8 +42,11 @@ def Recv(
 
 
 #  this function executes the primitive, when not under any transformation
-def mpi_recv_impl(*args, **kwargs):
-    return xla.apply_primitive(mpi_recv_impl, *args, **kwargs)
+def mpi_recv_impl(x, token, source, tag, comm, status):
+    comm = wrap_as_hashable(comm)
+    return xla.apply_primitive(
+        mpi_recv_p, x, token, source=source, tag=tag, comm=comm, status=status
+    )
 
 
 #  This function compiles the operation
@@ -49,6 +54,8 @@ def mpi_recv_xla_encode(c, x, token, source, tag, comm, status):
     from ..cython.mpi_xla_bridge import MPI_STATUS_IGNORE_ADDR
 
     warn_missing_omnistaging()
+
+    comm = unpack_hashable(comm)
 
     c = _unpack_builder(c)
     x_shape = c.GetShape(x)
@@ -66,7 +73,7 @@ def mpi_recv_xla_encode(c, x, token, source, tag, comm, status):
     if status is None:
         _status = MPI_STATUS_IGNORE_ADDR
     else:
-        _status = _MPI._addressof(status)
+        _status = to_mpi_ptr(status)
 
     operands = (
         _nitems,

--- a/mpi4jax/collective_ops/recv.py
+++ b/mpi4jax/collective_ops/recv.py
@@ -17,12 +17,14 @@ from ..utils import (
     dtype_ptr,
     wrap_as_hashable,
     unpack_hashable,
+    default_primitive_impl,
 )
 
 from ..warn import warn_missing_omnistaging
 
 # The Jax primitive
 mpi_recv_p = Primitive("recv_mpi")  # Create the primitive
+mpi_recv_impl = default_primitive_impl(mpi_recv_p)
 
 
 # This function applies the primitive to an AST
@@ -37,16 +39,8 @@ def Recv(
     if token is None:
         token = create_token(x)
 
-    out = mpi_recv_p.bind(x, token, source=source, tag=tag, comm=comm, status=status)
-    return out
-
-
-#  this function executes the primitive, when not under any transformation
-def mpi_recv_impl(x, token, source, tag, comm, status):
     comm = wrap_as_hashable(comm)
-    return xla.apply_primitive(
-        mpi_recv_p, x, token, source=source, tag=tag, comm=comm, status=status
-    )
+    return mpi_recv_p.bind(x, token, source=source, tag=tag, comm=comm, status=status)
 
 
 #  This function compiles the operation

--- a/mpi4jax/collective_ops/send.py
+++ b/mpi4jax/collective_ops/send.py
@@ -17,6 +17,7 @@ from ..utils import (
     dtype_ptr,
     wrap_as_hashable,
     unpack_hashable,
+    default_primitive_impl,
 )
 
 from ..warn import warn_missing_omnistaging
@@ -24,6 +25,7 @@ from ..warn import warn_missing_omnistaging
 
 # The Jax primitive
 mpi_send_p = Primitive("send_mpi")  # Create the primitive
+mpi_send_impl = default_primitive_impl(mpi_send_p)
 
 
 # This function applies the primitive to an AST
@@ -31,14 +33,8 @@ def Send(x, dest, tag=0, comm=_MPI.COMM_WORLD, token=None):
     if token is None:
         token = create_token(x)
 
-    out = mpi_send_p.bind(x, token, dest=dest, tag=tag, comm=comm)
-    return out
-
-
-#  this function executes the primitive, when not under any transformation
-def mpi_send_impl(x, token, dest, tag, comm):
     comm = wrap_as_hashable(comm)
-    return xla.apply_primitive(mpi_send_p, x, token, dest=dest, tag=tag, comm=comm)
+    return mpi_send_p.bind(x, token, dest=dest, tag=tag, comm=comm)
 
 
 #  This function compiles the operation

--- a/mpi4jax/collective_ops/send.py
+++ b/mpi4jax/collective_ops/send.py
@@ -99,7 +99,6 @@ def mpi_send_abstract_eval(xs, token, dest, tag, comm):
     return abstract_arrays.abstract_token
 
 
-# mpi_send_p.multiple_results = True
 mpi_send_p.def_impl(mpi_send_impl)
 mpi_send_p.def_abstract_eval(mpi_send_abstract_eval)
 

--- a/mpi4jax/collective_ops/send.py
+++ b/mpi4jax/collective_ops/send.py
@@ -15,6 +15,8 @@ from ..utils import (
     _constant_s32_scalar,
     _constant_u64_scalar,
     dtype_ptr,
+    wrap_as_hashable,
+    unpack_hashable,
 )
 
 from ..warn import warn_missing_omnistaging
@@ -34,13 +36,16 @@ def Send(x, dest, tag=0, comm=_MPI.COMM_WORLD, token=None):
 
 
 #  this function executes the primitive, when not under any transformation
-def mpi_send_impl(*args, **kwargs):
-    xla.apply_primitive(mpi_send_p, *args, **kwargs)
+def mpi_send_impl(x, token, dest, tag, comm):
+    comm = wrap_as_hashable(comm)
+    return xla.apply_primitive(mpi_send_p, x, token, dest=dest, tag=tag, comm=comm)
 
 
 #  This function compiles the operation
 def mpi_send_xla_encode(c, x, token, dest, tag, comm):
     warn_missing_omnistaging()
+
+    comm = unpack_hashable(comm)
 
     c = _unpack_builder(c)
     x_shape = c.GetShape(x)

--- a/mpi4jax/collective_ops/send.py
+++ b/mpi4jax/collective_ops/send.py
@@ -34,11 +34,8 @@ def Send(x, dest, tag=0, comm=_MPI.COMM_WORLD, token=None):
 
 
 #  this function executes the primitive, when not under any transformation
-def mpi_send_impl(x, token, dest, tag, comm):
-    # TODO: make this support gpus (use cupy?)
-    inpt = _np.asarray(x)
-    comm.Send(inpt, dest=dest, tag=tag)
-    return token
+def mpi_send_impl(*args, **kwargs):
+    xla.apply_primitive(mpi_send_p, *args, **kwargs)
 
 
 #  This function compiles the operation

--- a/mpi4jax/collective_ops/sendrecv.py
+++ b/mpi4jax/collective_ops/sendrecv.py
@@ -2,8 +2,7 @@ import numpy as _np
 
 from mpi4py import MPI as _MPI
 
-import jax.numpy as jnp
-from jax import abstract_arrays, device_put
+from jax import abstract_arrays
 from jax.lax import create_token
 from jax.core import Primitive
 from jax.lib import xla_client
@@ -53,38 +52,8 @@ def Sendrecv(
 
 
 #  this function executes the primitive, when not under any transformation
-def mpi_sendrecv_impl(
-    sendbuf, recvbuf, token, source, dest, sendtag, recvtag, comm, status
-):
-    # TODO: make this support gpus (use cupy?)
-    inarr = _np.asarray(sendbuf)
-    outarr = _np.empty_like(recvbuf)
-
-    comm.Sendrecv(
-        sendbuf=inarr,
-        dest=dest,
-        sendtag=sendtag,
-        recvbuf=outarr,
-        source=source,
-        recvtag=recvtag,
-        status=status,
-    )
-
-    if hasattr(recvbuf, "dtype"):
-        dt = recvbuf.dtype
-    else:
-        # probably a scalar
-        dt = _np.dtype(type(recvbuf))
-
-    res = jnp.array(outarr, dtype=dt)
-
-    # if it's a jax array and not a standard python array
-    if hasattr(recvbuf, "device_buffer"):
-        # put the result on the correct device if needed
-        if not (res.device_buffer.device() == recvbuf.device_buffer.device()):
-            res = device_put(res, device=recvbuf.device_buffer.device())
-
-    return res, token
+def mpi_sendrecv_impl(*args, **kwargs):
+    return xla.apply_primitive(mpi_sendrecv_p, *args, **kwargs)
 
 
 #  This function compiles the operation

--- a/mpi4jax/collective_ops/sendrecv.py
+++ b/mpi4jax/collective_ops/sendrecv.py
@@ -17,12 +17,14 @@ from ..utils import (
     dtype_ptr,
     wrap_as_hashable,
     unpack_hashable,
+    default_primitive_impl,
 )
 
 from ..warn import warn_missing_omnistaging
 
 # The Jax primitive
 mpi_sendrecv_p = Primitive("sendrecv_mpi")  # Create the primitive
+mpi_sendrecv_impl = default_primitive_impl(mpi_sendrecv_p)
 
 
 # This function applies the primitive to an AST
@@ -40,26 +42,8 @@ def Sendrecv(
     if token is None:
         token = create_token(sendbuf)
 
-    return mpi_sendrecv_p.bind(
-        sendbuf,
-        recvbuf,
-        token,
-        source=source,
-        dest=dest,
-        sendtag=sendtag,
-        recvtag=recvtag,
-        comm=comm,
-        status=status,
-    )
-
-
-#  this function executes the primitive, when not under any transformation
-def mpi_sendrecv_impl(
-    sendbuf, recvbuf, token, source, dest, sendtag, recvtag, comm, status
-):
     comm = wrap_as_hashable(comm)
-    return xla.apply_primitive(
-        mpi_sendrecv_p,
+    return mpi_sendrecv_p.bind(
         sendbuf,
         recvbuf,
         token,

--- a/mpi4jax/collective_ops/sendrecv.py
+++ b/mpi4jax/collective_ops/sendrecv.py
@@ -15,6 +15,8 @@ from ..utils import (
     _constant_s32_scalar,
     _constant_u64_scalar,
     dtype_ptr,
+    wrap_as_hashable,
+    unpack_hashable,
 )
 
 from ..warn import warn_missing_omnistaging
@@ -52,8 +54,22 @@ def Sendrecv(
 
 
 #  this function executes the primitive, when not under any transformation
-def mpi_sendrecv_impl(*args, **kwargs):
-    return xla.apply_primitive(mpi_sendrecv_p, *args, **kwargs)
+def mpi_sendrecv_impl(
+    sendbuf, recvbuf, token, source, dest, sendtag, recvtag, comm, status
+):
+    comm = wrap_as_hashable(comm)
+    return xla.apply_primitive(
+        mpi_sendrecv_p,
+        sendbuf,
+        recvbuf,
+        token,
+        source=source,
+        dest=dest,
+        sendtag=sendtag,
+        recvtag=recvtag,
+        comm=comm,
+        status=status,
+    )
 
 
 #  This function compiles the operation
@@ -63,6 +79,8 @@ def mpi_sendrecv_xla_encode(
     from ..cython.mpi_xla_bridge import MPI_STATUS_IGNORE_ADDR
 
     warn_missing_omnistaging()
+
+    comm = unpack_hashable(comm)
 
     c = _unpack_builder(c)
 
@@ -92,7 +110,7 @@ def mpi_sendrecv_xla_encode(
     if status is None:
         _status = MPI_STATUS_IGNORE_ADDR
     else:
-        _status = _MPI._addressof(status)
+        _status = to_mpi_ptr(status)
 
     operands = (
         _send_nitems,

--- a/mpi4jax/utils.py
+++ b/mpi4jax/utils.py
@@ -1,5 +1,4 @@
 import numpy as _np
-import ctypes
 
 from mpi4py import MPI as _MPI
 
@@ -14,31 +13,13 @@ def to_mpi_ptr(mpi_obj):
 
     Returns the ptr to the underlying C mpi object
     """
-    return _np.uint64(_MPI._handleof(mpi_obj))
+    try:
+        addr = _MPI._handleof(mpi_obj)
+    except NotImplementedError:
+        # some objects like Status only work with addressof
+        addr = _MPI._addressof(mpi_obj)
 
-
-def MPIComm_from_ptr(ptr):
-    """
-    MPIComm_from_ptr(ptr)
-
-    Constructs a MPI Comm object from a pointer
-    """
-    comm = _MPI.Comm()
-    comm_ptr = ctypes.c_void_p.from_address(_MPI._addressof(comm))
-    comm_ptr.value = int(ptr)
-    return comm
-
-
-def MPIOp_from_ptr(ptr):
-    """
-    MPIOp_from_ptr(ptr)
-
-    Constructs a MPI Op object from a pointer
-    """
-    op = _MPI.Op()
-    op_ptr = ctypes.c_void_p.from_address(_MPI._addressof(op))
-    op_ptr.value = int(ptr)
-    return op
+    return _np.uint64(addr)
 
 
 def dtype_ptr(dtype):
@@ -63,7 +44,31 @@ def dtype_ptr(dtype):
     return _dtype
 
 
+# Helpers to make MPI objects hashable
+
+
+class HashableMPIType:
+    def __init__(self, obj):
+        self.wrapped = obj
+
+    def __hash__(self):
+        return int(to_mpi_ptr(self.wrapped))
+
+
+def wrap_as_hashable(obj):
+    return HashableMPIType(obj)
+
+
+def unpack_hashable(obj):
+    if isinstance(obj, HashableMPIType):
+        return obj.wrapped
+
+    return obj
+
+
 # Helper functions
+
+
 def _constant_s32_scalar(c, x):
     return _ops.Constant(c, _np.int32(x))
 

--- a/mpi4jax/utils.py
+++ b/mpi4jax/utils.py
@@ -1,10 +1,16 @@
+import functools
 import numpy as _np
 
 from mpi4py import MPI as _MPI
 
 from jax.lib import xla_client
+from jax.interpreters import xla
 
 _ops = xla_client.ops
+
+
+def default_primitive_impl(primitive):
+    return functools.partial(xla.apply_primitive, primitive)
 
 
 def to_mpi_ptr(mpi_obj):

--- a/mpi4jax/utils.py
+++ b/mpi4jax/utils.py
@@ -62,6 +62,9 @@ class HashableMPIType:
 
 
 def wrap_as_hashable(obj):
+    if isinstance(obj, HashableMPIType):
+        return obj
+
     return HashableMPIType(obj)
 
 

